### PR TITLE
Initial commit of P0 Markdown features for Rooster

### PR DESCRIPTION
**Pull request by ctuck04**
The Markdown features include:
-- Bold
-- Underline
-- Italic
-- Strikethrough

### CODE REVIEW

File: packages/roosterjs-editor-core/lib/interfaces/ContentEditFeature.ts [hunk 1]
```diff
@@ -14,14 +14,18 @@ export const enum Keys {
...
     I = 73,
     U = 85,
     Y = 89,
     Z = 90,
+    PLUS = 187,
     COMMA = 188,
+    UNDERSCORE = 189,
     PERIOD = 190,
     FORWARDSLASH = 191,
+    SQUIGGLY = 192,
```
Comment (Lego6245) [inline packages/roosterjs-editor-core/lib/interfaces/ContentEditFeature.ts:28]: nit: should be tildehttps://en.wikipedia.org/wiki/Tilde

File: packages/roosterjs-editor-plugins/lib/ContentEdit/ContentEditFeatures.ts [hunk 2]
```diff
@@ -131,5 +139,9 @@ export function getDefaultContentEditFeatures(): ContentEditFeatures {
         noCycleCursorMove: Browser.isChrome,
         smartOrderedList: false,
         smartOrderedListStyles: ['lower-alpha', 'lower-roman', 'decimal'],
+        bold: true,
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/ContentEditFeatures.ts:142]: we probably would not include these in the default enabled set, and instead pass them in to explicitly enable them

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/autoLinkFeatures.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ import {
     Editor,
     cacheGetEventData,
     cacheGetContentSearcher,
-    clearContentSearcherCache,
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/autoLinkFeatures.ts:15]: nit: clean this up

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/autoLinkFeatures.ts [hunk 4]
```diff
@@ -114,7 +114,7 @@ function autoLink(event: PluginEvent, editor: Editor) {
             replaceWithNode(editor, linkData.originalUrl, anchor, false /* exactMatch */, searcher);
 
             // The content at cursor has changed. Should also clear the cursor data cache
-            clearContentSearcherCache(event);
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/autoLinkFeatures.ts:117]: nit: revert?

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        anchorNode.textContent.lastIndexOf(triggerChar) == 0 ||
+        anchorNode.textContent.lastIndexOf(triggerChar) == 1
+    ) {
+        genericTriggerCharacterIndex = anchorNode.textContent.lastIndexOf(triggerChar);
+    } else if (anchorNode.textContent.lastIndexOf(' ' + triggerChar) >= 0) {
+        genericTriggerCharacterIndex = anchorNode.textContent.lastIndexOf(' ' + triggerChar);
+    }
+
+    let firstWhiteSpaceChecker = anchorNode.textContent.charAt(
+        genericTriggerCharacterIndex + stringAdjust
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:31]: why is stringAdjust a variable here when it's being used as a const

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    event: PluginKeyboardEvent,
+    editor: Editor
+) => {
+    let selObj = window.getSelection();
+    let anchorNode = selObj.anchorNode;
+    let genericTriggerCharacterIndex = anchorNode.textContent.lastIndexOf(' ' + triggerChar);
+    let alterationRequired = false;
+    let stringAdjust = 2;
+
+    if (
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:21]: compute this value once, and check it each time. goes for both of these conditionals

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+) => {
+    let selObj = window.getSelection();
+    let anchorNode = selObj.anchorNode;
+    let stringToBeAltered = new Range();
+    let genericTriggerCharacterRange = new Range();
+    let stringAdjust = 1;
+    let scenario = 'N/A';
+    let genericTriggerCharacterIndex;
+
+    if (
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:53]: could probably pull this out into its own function

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    } else if (anchorNode.textContent.lastIndexOf(' ' + triggerChar) >= 0) {
+        genericTriggerCharacterIndex = anchorNode.textContent.lastIndexOf(' ' + triggerChar);
+        scenario = 'mid';
+    }
+
+    if (scenario == 'begin') {
+        stringToBeAltered.setStart(anchorNode, genericTriggerCharacterIndex + stringAdjust);
+        stringToBeAltered.setEnd(anchorNode, anchorNode.textContent.length);
+
+        genericTriggerCharacterRange.setStart(anchorNode, genericTriggerCharacterIndex);
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:68]: `anchorNode.textContent.replaceAll` might be simpler here and below

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        anchorNode.textContent.lastIndexOf(triggerChar) == 1
+    ) {
+        genericTriggerCharacterIndex = anchorNode.textContent.lastIndexOf(triggerChar);
+        scenario = 'begin';
+    } else if (anchorNode.textContent.lastIndexOf(' ' + triggerChar) >= 0) {
+        genericTriggerCharacterIndex = anchorNode.textContent.lastIndexOf(' ' + triggerChar);
+        scenario = 'mid';
+    }
+
+    if (scenario == 'begin') {
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:64]: could move this if block into the above if block

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    handleEvent: applyStyle('+', 'u'),
+};
+
+export const Strikethrough: ContentEditFeature = {
+    keys: [Keys.SQUIGGLY],
+    shouldHandleEvent: findTriggerCharactersAndCheckBetween('~'),
+    handleEvent: applyStyle('~', 'strike'),
+};
+
+// 6/4/2020 1:16 PM PDT
```
Comment (Lego6245) [inline packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:125]: nit: clean this up

File: packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts (hunk 8 above)
Comment (ctuck04) [reply packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:68]: I'm getting "Property 'replaceAll' does not exist on type 'string'."

Comment (Lego6245) [reply packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:68]: whoops, `replaceAll` isn't in all browsers: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replaceAll

Comment (Lego6245) [reply packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:68]: `replace` should work just as well (replace and replace all both do the same thing pretty much)

Comment (Lego6245) [reply packages/roosterjs-editor-plugins/lib/ContentEdit/features/markdownFeatures.ts:68]: you could use `slice` here alternatively, or leave as is. not a big deal either way, just think it's cleaner than making a range and deleting its contents

**Pull request closed without merging**